backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,318 @@
+        """
+        logger.info(f"Fetching fire events in {state} in year {year}.")
+        current_year = datetime.datetime.now().date().year
+        url_year = "current_year" if current_year == year else year
+        fire_events_page_in_state = FireCrawler._get_url(f"{FireCrawler.BASE_DIR}{url_year}_fire_data/{state}",
+                                                         f"{state} State of year {year}",
+                                                         "_extract_fire_events_in_state()")
+        fire_names_in_this_state = FireCrawler.RE_EXTRACT_FIRE_EVENTS_IN_STATE.findall(fire_events_page_in_state)
+        logger.info(f"Finished fetching fire events in {state} in year {year}.")
+        return list(map(lambda single_fire_name: FireEvent(year, state, single_fire_name), fire_names_in_this_state))
+
+    @staticmethod
+    def _filter_out_invalid_fire_events(fires: List[FireEvent]):
+        """
+        Find those invalid FireEvent objects and remove them from the input list
+        :param fires: list of FireEvent objects.
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        for fire_event in fires[:]:
+            if fire_event.is_valid():
+                fires.remove(fire_event)
+
+    def extract_all_fires(self, start_year: int) -> List[FireEvent]:
+        """
+        Extracts all fires on the website and return.
+        The returned value is a list of tuples:(year:int, state:string, fire's name in its url)
+        :param start_year: int, start year for the crawler. Record before the start year will be deleted.
+                e.g. 2015
+        :return: list of FireEvent objects
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        logger.info("Attempting to get all links...")
+        current_year = datetime.datetime.now().date().year
+        # fetch all year nodes from the website, may encounter errors
+        # add a try-except block, and while loop to keep trying
+        # if attempts > 3 raise an exception
+        try:
+            main_page = self._get_url(FireCrawler.BASE_DIR, "main_page", "extract_all_fires()")
+        except requests.exceptions.RequestException:
+            logger.error("Error: Cannot fetch main_page in function extract_all_fires()")
+            raise CannotCrawlException("Error: Cannot fetch main_page in function extract_all_fires")
+            # retry exception
+        else:
+            # the last node is "historical_fire_data" and we don't want it
+            sub_links_of_each_year = FireCrawler.RE_EXTRACT_ALL_FIRES.findall(main_page)[:-1]
+            # sub_links_of_each_year are now fire data of a certain year or current year
+            # e.g. '2010_fire_data', 'current_year_fire_data'
+            fires = []
+            for sub_link_of_each_year in sub_links_of_each_year:
```

Comment:
not a good practice to put plural form in the middle of variable name. it is hard to see the difference!